### ISSUE TITLE
inventory should be sorted before items remaining is displayed

### ISSUE THREAD

**Issue by angband-import** [angband/angband#1557]
**Reported by anonymous on 11 Dec 2009 15:22 UTC**
When identifying a new item with a scroll, the number of scrolls remaining is listed after it's read, with the current inventory slot.  The inventory is then sorted, potentially changing this slot without informing the player of the new slot.

**Comment by angband-import** [comment]
**Modified by magnate on 23 Jan 2010 09:00 UTC**

**Comment by angband-import** [comment]
**Comment by magnate on 21 Nov 2010 20:03 UTC**
Punting to 3.3: non-urgent bug or change.

**Comment by angband-import** [comment]
**Comment by magnate on 18 Dec 2010 18:23 UTC**
Confirmed that this still happens as of 3.2rc1.

**Comment by angband-import** [comment]
**Comment by magnate on 22 May 2011 19:14 UTC**
Punting in accordance with new milestone policy (that any other milestone is only set once someone is actually working on the ticket).

**Comment by angband-import** [comment]
**Comment by takkaria on 27 Apr 2013 23:13 UTC**
Assigning still-open bugs from Future to 3.5 for fixing.

**Comment by angband-import** [comment]
**Comment by takkaria on 10 Oct 2013 18:38 UTC**
Fixed with the pull from https://github.com/angband/angband/pull/274 .

**Comment by angband-import** [comment]
**Comment by PowerWyrm on 17 Oct 2013 11:55 UTC**
This change seems to have introduced some new bugs:

- activatable equipped items instantly recharge
- reading a scroll of identify from the ground and identifying an equipped item gives a message saying that the identified item is on the ground

This is probably due to a mismatch in do_cmd_use() between the original object and its copy in the code.

**Comment by angband-import** [comment]
**Modified by molybdenum on 18 Oct 2013 16:01 UTC**

**Comment by angband-import** [comment]
**Comment by PowerWyrm on 23 Oct 2013 14:01 UTC**
While we're at it, maybe the slot info shouldn't be displayed when using up the last item: "You have no more foo." instead of "You have no more foo (slot)."

**Comment by angband-import** [comment]
**Comment by PowerWyrm on 24 Oct 2013 15:21 UTC**
Ok now do_ident_item() reorders the pack before displaying the updated message, but this means that the parameters of the function are no longer valid when the function returns (mainly the current object). This means that any call of do_ident_item() should be isolated and further references to the manipulated object should be avoided.

Looking at the code, do_ident_item() is called five times:
- in sense_inventory(): called inside a loop, potentially dangerous since a reordering would completely break the loop index; this is really tricky to fix since there is no way to tell how many times the pack will be reordered while sensing it; I don't see any other way than tagging every object in the pack with a flag saying that the object has been processed in the loop and reset the loop each time the pack is reordered; something like

```
i = 0;
while (i < ALL_INVEN_TOTAL)
{
  object_type *o_ptr = &p_ptr->inventory[i];

  ...

  /* It has already been processed */
  if (o_ptr->tagged) continue;

  if (...)
  {
    do_ident_item(i, o_ptr);
    i = 0;
    continue;
  }

  ...

  o_ptr->tagged = TRUE;
  i++;
}
```

- in inven_carry(): two successive calls with the same object parameter, potentially dangerous since the parameter may have changed; easily fixed with an "else"

```
if (player_has(PF_KNOW_MUSHROOM) && j_ptr->tval == TV_FOOD)
{
  do_ident_item(i, j_ptr);
  msg("Mushrooms for breakfast!");
}
else if (player_has(PF_KNOW_ZAPPER) &&
  (j_ptr->tval == TV_WAND || j_ptr->tval == TV_STAFF))
{
  do_ident_item(i, j_ptr);
}
```

- in identify_pack(): called inside a loop, potentially dangerous since a reordering would completely break the loop index; partially fixed already with the "i--" line; replacing that by "i = 0" should ensure that everything is correctly described when identifying everything in the pack (see *enlightenment* effect)

- in ident_spell(): isolated call, should be fixed by recent commits

**Comment by angband-import** [comment]
**Comment by takkaria on 28 Oct 2013 22:48 UTC**
Fixed in c9e89af, pullreq https://github.com/angband/angband/pull/279
